doc/Moving-to-single-app-project-structure.md

```diff
@@ -0,0 +1,160 @@
+- `application:start`
+- `application:stop`
+- `application:ensure_all_started`
+- `code:priv_dir`
+
+Those function are also called via RPC from big test suites. A good method to find them all is to
+search using `application.*ej` regex.
+
+Files which require changes:
+- `include/ejabberd.hrl`
+- `src/ejabberd.erl`
+- `src/ejabberd_app.erl`
+- `src/ejabberd_config.erl`
+- `src/mod_version.erl`
+- `src/translate.erl`
+- `test/ejabberd_config_SUITE.erl`
+- `test/ejabberd_helper.erl`
+- `test/ejabberd_listener_SUITE.erl`
+- `test/zlib_driver_SUITE.erl`
+- `test.disabled/ejabberd_tests/README.md`
+- `test.disabled/ejabberd_tests/run_common_test.erl`
+- `test.disabled/ejabberd_tests/tests/conf_reload_SUITE.erl`
+- `test.disabled/ejabberd_tests/tests/connect_SUITE.erl`
+- `test.disabled/ejabberd_tests/tests/reload_helper.erl`
+- `tools/extract_translations/extract_translations.erl`
+- `tools/xep_tool/xep_tool.escript`
+- `src/mongoose_cluster.erl` (2 calls to `when_app_stopped/1` function)
+
+We also need to remove remaining occurences of `apps/ejabberd` directory structure. Some of the
+files are scripts assembling the path programatically - the trick is to search for the `apps/` keyword.
+
+Affected files:
+- `CONTRIBUTING.md`
+- `Makefile`
+- `elvis.config`
+- `src/amp_strategy.erl`
+- `src/mod_mam_odbc_arch.erl`
+- `test.disabled/ejabberd_tests/Makefile`
+- `test.disabled/ejabberd_tests/run_common_test.erl`
+- `test.disabled/ejabberd_tests/tests/vcard_SUITE.erl`
+- `tools/cd_tools/provision_mysql.sh`
+- `tools/cd_tools/provision_pgsql.sh`
+- `tools/fill_roster.erl`
+- `tools/privacy.erl`
+- `tools/travis-setup-db.sh`
+- `tools/travis-test.sh`
+- `tools/travis-upload-to-s3.sh`
+- `tools/zabbix_add_counters.py`
+
+## Moving a MongooseIM's fork to a new structure
+
+This section describes how to move customized fork of MongooseIM to a new structure.
+
+1. Move all directories from `apps/ejabberd` directory to the repository root, e.g. `apps/ejabberd/src/`
+   should be moved to just `src/`. Be sure to move all the important files, because `apps/` directory
+   will be wiped out completely.
+
+2. Make all the changes described in `Move tools files` section.  If you've made any changes to
+   `apps/ejabberd/rebar.config` or `apps/ejabberd/rebar.config/script`, be sure to merge these
+   with `rebar.config` and `rebar.config.script` in the repository root. Remove `apps/` directory.
+
+3. Make all the changes described in `Application resource files` section.
+
+4. Make all the changes described in `Includes` section. Make sure that no include directives
+   in your source code (in `src/` and `test/`) look like `-include_lib("ejabberd/include/<header_file>")`.
+   They should now refer to bare header file names, i.e. `-include(<header_file>)`.
+
+5. Make all the changes described in `Release scripts` section. Make sure that all customizations
+   made to release scripts and release process are aware of new directory structure (i.e. none
```

Comment:
made to release scripts and release process are aware of the new directory structure (i.e. none